src/ucp/tag/rndv.c

```diff
@@ -610,6 +617,205 @@ static void ucp_rndv_req_send_rma_get(ucp_request_t *rndv_req, ucp_request_t *rr
+    }
+
+    ucp_request_put(freq);
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_rndv_init_mem_type_frag_req(ucp_worker_h worker, ucp_request_t *freq, int rndv_op,
+                                uct_completion_callback_t comp_cb, ucp_mem_desc_t *mdesc,
+                                ucs_memory_type_t mem_type, size_t length,
+                                uct_pending_callback_t uct_func)
+{
+    ucp_ep_h mem_type_ep;
+    ucp_md_index_t md_index;
+    ucp_lane_index_t mem_type_rma_lane;
+
+    ucp_request_send_state_init(freq, ucp_dt_make_contig(1), 0);
+    ucp_request_send_state_reset(freq, comp_cb, rndv_op);
+
+    freq->send.buffer   = mdesc + 1;
+    freq->send.length   = length;
+    freq->send.datatype = ucp_dt_make_contig(1);
+    freq->send.mem_type = mem_type;
+    freq->send.mdesc    = mdesc;
+    freq->send.uct.func = uct_func;
+
+    if (mem_type != UCS_MEMORY_TYPE_HOST) {
+        mem_type_ep       = worker->mem_type_ep[mem_type];
+        mem_type_rma_lane = ucp_ep_config(mem_type_ep)->key.rma_bw_lanes[0];
+        md_index          = ucp_ep_md_index(mem_type_ep, mem_type_rma_lane);
+        ucs_assert(mem_type_rma_lane != UCP_NULL_LANE);
+
+        freq->send.lane                       = mem_type_rma_lane;
+        freq->send.ep                         = mem_type_ep;
+        freq->send.state.dt.dt.contig.memh[0] = ucp_memh2uct(mdesc->memh, md_index);
+        freq->send.state.dt.dt.contig.md_map  = UCS_BIT(md_index);
+    }
+}
+
+static void
+ucp_rndv_recv_frag_put_mem_type(ucp_request_t *rreq, ucp_request_t *rndv_req,
+                                ucp_request_t *freq, ucp_mem_desc_t *mdesc,
+                                size_t length, size_t offset)
+{
+
+    ucs_assert_always(!UCP_MEM_IS_ACCESSIBLE_FROM_CPU(rreq->recv.mem_type));
+
+    /* PUT on memtype endpoint to stage from
+     * frag recv buffer to memtype recv buffer
+     */
+
+    ucp_rndv_init_mem_type_frag_req(rreq->recv.worker, freq, UCP_REQUEST_SEND_PROTO_RNDV_PUT,
+                                    ucp_rndv_recv_frag_put_completion, mdesc, rreq->recv.mem_type,
+                                    length, ucp_rndv_progress_rma_put_zcopy);
+
+    freq->send.rndv_put.sreq           = rreq;
+    freq->send.rndv_put.rkey           = NULL;
+    freq->send.rndv_put.remote_request = (uintptr_t) rndv_req;
+    freq->send.rndv_put.remote_address =
+                    (uint64_t)UCS_PTR_BYTE_OFFSET(rreq->recv.buffer, offset);
+
+    ucp_request_send(freq, 0);
+}
+
+static ucs_status_t
+ucp_rndv_send_frag_get_mem_type(ucp_request_t *sreq, uintptr_t rreq_ptr,
+                                size_t length, uint64_t remote_address,
+                                ucs_memory_type_t remote_mem_type, ucp_rkey_h rkey,
+                                uct_completion_callback_t comp_cb)
+{
+    ucp_worker_h worker = sreq->send.ep->worker;
+    ucp_request_t *freq;
+    ucp_mem_desc_t *mdesc;
+
+    /* GET on memtype endpoint to stage from
```

Comment:
it is not the only case, right?
we also use it for getting data from the peer (no necessarily from memtype ep)

Reply:
fixed the vebose